metrics/sinks/influxdb/influxdb_historical.go

```diff
@@ -0,0 +1,550 @@
+
+// Kube pod and namespace names are limitted to [a-zA-Z0-9-.], while docker also allows
+// underscores, so only allow these those characters.  When Influx actually supports bound
+// paramaters, this will be less necessary.
+var nameAllowedChars = regexp.MustCompile("^[a-zA-Z0-9_.-]+$")
+
+// sanitizeKey errors out if invalid characters are found in the key, since InfluxDB does not widely
+// support bound parameters yet (see https://github.com/influxdata/influxdb/pull/6634) and we need to
+// sanitize our inputs.
+func (sink *influxdbSink) checkSanitizedKey(key *core.HistoricalKey) error {
+	if key.NodeName != "" && !nameAllowedChars.MatchString(key.NodeName) {
+		return fmt.Errorf("Invalid node name %q", key.NodeName)
+	}
+
+	if key.NamespaceName != "" && !nameAllowedChars.MatchString(key.NamespaceName) {
+		return fmt.Errorf("Invalid namespace name %q", key.NamespaceName)
+	}
+
+	if key.PodName != "" && !nameAllowedChars.MatchString(key.PodName) {
+		return fmt.Errorf("Invalid pod name %q", key.PodName)
+	}
+
+	// NB: this prevents access to some of the free containers with slashes in their name
+	// (e.g. system.slice/foo.bar), but the Heapster API seems to choke on the slashes anyway
+	if key.ContainerName != "" && !nameAllowedChars.MatchString(key.ContainerName) {
+		return fmt.Errorf("Invalid container name %q", key.ContainerName)
+	}
+
+	if key.PodId != "" && !nameAllowedChars.MatchString(key.PodId) {
+		return fmt.Errorf("Invalid pod id %q", key.PodId)
+	}
+
+	return nil
+}
+
+// aggregationFunc converts an aggregation name into the equivalent call to an InfluxQL
+// aggregation function
+func (sink *influxdbSink) aggregationFunc(aggregationName core.AggregationType, fieldName string) string {
+	switch aggregationName {
+	case core.AggregationTypeAverage:
+		return fmt.Sprintf("MEAN(%q)", fieldName)
+	case core.AggregationTypeMaximum:
+		return fmt.Sprintf("MAX(%q)", fieldName)
+	case core.AggregationTypeMinimum:
+		return fmt.Sprintf("MIN(%q)", fieldName)
+	case core.AggregationTypeMedian:
+		return fmt.Sprintf("MEDIAN(%q)", fieldName)
+	case core.AggregationTypeCount:
+		return fmt.Sprintf("COUNT(%q)", fieldName)
+	case core.AggregationTypePercentile50:
+		return fmt.Sprintf("PERCENTILE(%q, 50)", fieldName)
+	case core.AggregationTypePercentile95:
+		return fmt.Sprintf("PERCENTILE(%q, 95)", fieldName)
+	case core.AggregationTypePercentile99:
+		return fmt.Sprintf("PERCENTILE(%q, 99)", fieldName)
+	}
+
+	// This should have been checked by the API level, so something's seriously wrong here
+	panic(fmt.Sprintf("Unknown aggregation type %q", aggregationName))
+}
+
+// keyToSelector converts a HistoricalKey to a InfluxQL predicate
+func (sink *influxdbSink) keyToSelector(key core.HistoricalKey) string {
+	typeSel := fmt.Sprintf("type = '%s'", key.ObjectType)
+	switch key.ObjectType {
+	case core.MetricSetTypeNode:
+		return fmt.Sprintf("%s AND %s = '%s'", typeSel, core.LabelNodename.Key, key.NodeName)
```

Comment:
Key values are checked in `checkSanitizedKey` above, and an error is thrown if anything besides `/^[a-zA-Z0-9_.-]+$/` is found.